File: LeetCode/Medium/0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode node = new ListNode();
        ListNode curr = node;
        int k = left;

        while (k > 1) {
            curr.next = new ListNode(head.val);
            curr = curr.next;
            head = head.next;
            k--;
        }

        Deque<Integer> dq = new ArrayDeque<>();
        
        while (head != null && left <= right) {
            dq.offer(head.val);
            head = head.next;
            left++;
        }

        while (!dq.isEmpty()) {
            curr.next = new ListNode(dq.pollLast());
            curr = curr.next;
        }

        while (head != null) {
            curr.next = new ListNode(head.val);
            curr = curr.next;
            head = head.next;
        }

        return node.next;
    }
}

File: LeetCode/Medium/0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun reverseBetween(head: ListNode?, left: Int, right: Int): ListNode? {
        val node = ListNode(0)
        var curr = node
        var h = head
        var k = 1

        while (h != null && k < left) {
            curr.next = ListNode(h.`val`)
            curr = curr.next
            h = h.next
            k++
        }

        val dq = ArrayDeque<Int>()
        
        while (h != null && k <= right) {
            dq.addLast(h.`val`)
            h = h.next
            k++
        }

        while (dq.isNotEmpty()) {
            curr.next = ListNode(dq.removeLast())
            curr = curr.next
        }

        while (h != null) {
            curr.next = ListNode(h.`val`)
            curr = curr.next
            h = h.next
        }

        return node.next
    }
}